### PR TITLE
JAMES-2687 Refactor InMemoryIntegrationResources

### PR DESCRIPTION
**Pull request by chibenwa**
- Merge InMemoryIntegrationResources & InMemoryIntegrationResources.Resources
 - Specialize ManagerTestResources into a data provisionner and rely on InMemoryIntegrationResources
 - InMemoryIntegrationResources should use a single factory and avoid using static factory methods
 - Clean a bit the hell in memory mailbox test
 - Rely more on InMemoryIntegrationResources instead of MailboxManager constructor (now that we can)

### CODE REVIEW

**Comment by chibenwa** [comment]
```
[febf734bae6c3be2971bf2e6bfd6a8462c9a64ae] 08:07:34.874 [WARN ] c.d.d.c.Cluster - Re-preparing already prepared query is generally an anti-pattern and will likely affect performance. Consider preparing the statement only once. Query='SELECT * FROM blobParts WHERE id=:id AND chunkNumber=:chunkNumber;'
[febf734bae6c3be2971bf2e6bfd6a8462c9a64ae] Mar 15, 2019 8:07:47 AM org.junit.platform.launcher.core.DefaultLauncher handleThrowable
[febf734bae6c3be2971bf2e6bfd6a8462c9a64ae] WARNING: TestEngine with ID 'junit-jupiter' failed to execute tests
[febf734bae6c3be2971bf2e6bfd6a8462c9a64ae] java.lang.OutOfMemoryError: Java heap space
```

test this please

**Comment by chibenwa** [comment]
```
org.apache.james.mailbox.events.CassandraEventDeadLettersTest
[febf734bae6c3be2971bf2e6bfd6a8462c9a64ae] [ERROR] storeShouldKeepConsistencyWhenConcurrentStore  Time elapsed: 2.043 s  <<< FAILURE!
```

unlucky

test this please

File: mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/manager/InMemoryIntegrationResources.java [hunk 1]
```diff
@@ -107,6 +113,116 @@ public InMemoryPerUserMaxQuotaManager getMaxQuotaManager() {
         }
     }
 
+    public static class Factory {
```
Comment (trantienduchn) [inline mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/manager/InMemoryIntegrationResources.java]: It looks like a builder!

Comment (chibenwa) [reply mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/manager/InMemoryIntegrationResources.java]: True

File: mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/manager/InMemoryIntegrationResources.java [hunk 2]
```diff
@@ -107,6 +113,116 @@ public InMemoryPerUserMaxQuotaManager getMaxQuotaManager() {
...
+            this.groupMembershipResolver = Optional.empty();
+            this.authenticator = Optional.empty();
+            this.authorizator = Optional.empty();
+            this.eventBus = Optional.empty();
+            this.limitAnnotationCount = Optional.empty();
+            this.limitAnnotationSize = Optional.empty();
+            this.preDeletionHooksInstanciators = ImmutableSet.builder();
+        }
+
+        public Factory withAuthenticator(Authenticator authenticator) {
```
Comment (mbaechler) [inline mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/manager/InMemoryIntegrationResources.java]: could you remove the `with` prefix for consistency with the current codebase?

File: mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/manager/InMemoryIntegrationResources.java [hunk 3]
```diff
@@ -107,6 +113,116 @@ public InMemoryPerUserMaxQuotaManager getMaxQuotaManager() {
...
+                .collect(Guavate.toImmutableList()));
+            return this;
+        }
+
+        public Factory withPreDeletionHook(BiFunction<SessionProvider, InMemoryMailboxSessionMapperFactory, PreDeletionHook> preDeletionHook) {
+            this.preDeletionHooksInstanciators.add(preDeletionHook);
+            return this;
+        }
+
+        private BiFunction<SessionProvider, InMemoryMailboxSessionMapperFactory, PreDeletionHook> toInstanciator(PreDeletionHook preDeletionHook) {
```
Comment (mbaechler) [inline mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/manager/InMemoryIntegrationResources.java]: toFactory?

File: mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/manager/InMemoryIntegrationResources.java [hunk 4]
```diff
@@ -107,6 +113,116 @@ public InMemoryPerUserMaxQuotaManager getMaxQuotaManager() {
...
+            return (a, b) -> preDeletionHook;
+        }
+
+        public Resources create() {
+            InMemoryMailboxSessionMapperFactory mailboxSessionMapperFactory = new InMemoryMailboxSessionMapperFactory();
+            EventBus eventBus = this.eventBus.orElseGet(() -> new InVMEventBus(new InVmEventDelivery(new NoopMetricFactory())));
```
Comment (mbaechler) [inline mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/manager/InMemoryIntegrationResources.java]: isn't it dangerous to have default implementations for a test env? I mean, you can very well forget to provide an EventBus and have green tests when you think you were testing Rabbit EventBus.
What about making a staged builder with Enum values to make the pain as small as possible while still be descriptive about we run?

File: mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/manager/InMemoryIntegrationResources.java [hunk 5]
```diff
@@ -111,6 +113,17 @@ public StoreRightManager getStoreRightManager() {
         public InMemoryPerUserMaxQuotaManager getMaxQuotaManager() {
             return maxQuotaManager;
         }
+
+        public MessageIdManager createMessageIdManager() {
```
Comment (mbaechler) [inline mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/manager/InMemoryIntegrationResources.java]: it looks like a default choice, there may be several MessageIdManager, how can you decide which is the right one?

File: mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/manager/InMemoryIntegrationResources.java [hunk 6]
```diff
@@ -156,12 +168,16 @@ public InMemoryIntegrationResources create() {
...
                 quotaComponents,
                 index,
-                createHooks(sessionProvider, mailboxSessionMapperFactory));
+                createHooks(preInstanciationStage));
 
             eventBus.register(listeningCurrentQuotaUpdater);
             eventBus.register(new MailboxAnnotationListener(mailboxSessionMapperFactory, sessionProvider));
 
-            return new InMemoryIntegrationResources(manager, storeRightManager, new InMemoryMessageId.Factory(), currentQuotaManager, quotaRootResolver, maxQuotaManager, quotaManager);
+            if (index instanceof MailboxListener.GroupMailboxListener) {
```
Comment (mbaechler) [inline mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/manager/InMemoryIntegrationResources.java]: why?

Comment (chibenwa) [reply mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/manager/InMemoryIntegrationResources.java]: This was done in order to rely on IntegrationResources for Search index tests.

I did not come up with better design so I can very well drop this part of the refactoring.

Technical explanation:

You should not loose the type of the search index.

Maybe it should be a caller responsibility Akka:

```
InMemoryIntegrationResources resources = new InMemoryIntegrationResources.Factory()
            .withSearchIndex(preInstanciationStage -> new ElasticSearchListeningMessageSearchIndex(/* ... */)
            .create();

resources.getMailboxManager().getEventBus().register((MailboxListener.GroupMailboxListener) resources.getMailboxManager().getMessageSearchIndex());
```

WDYT ?

File: mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/manager/InMemoryIntegrationResources.java (hunk 5 above)
Comment (chibenwa) [reply mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/manager/InMemoryIntegrationResources.java]: I rewrote this, now this method is turned into:

```
    public MessageIdManager getMessageIdManager() {
        return storeMessageIdManager;
    }
```

BTW right now there is only one implementation for MessageIdManager which IMO makes a possible choice useless.

When needed we can very well add a method in the factory for allowing customizing this.

File: mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/manager/InMemoryIntegrationResources.java (hunk 4 above)
Comment (chibenwa) [reply mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/manager/InMemoryIntegrationResources.java]: > isn't it dangerous to have default implementations for a test env?

IMO it is safe to expect from an `InMemoryIntegrationResources` to be based on an `InVMEventBus`

I don't see why this step would be special here. Why not these alays ask for the search index in the builder? After all we are not sure this is the scanning one. Why not ask for the explicit list of preDeletionHooks? How about the AnnotationLimits? 

If it's to duplicate everywhere:

````
new Factory()
   .preProvisionnedFakeAuthenticator()
   .fakeAuthorizator()
   .inVmEventBus()
   .defaultAnnotationLimits()
   .storeQuotaManager()
   .defaultMessageParser()
   .scanningSearchIndex()
   .noPreDeletionHooks()
   .create();
```

Isn't that a bit long?Are we sure we do want this?

File: mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/manager/InMemoryIntegrationResources.java (hunk 6 above)
Comment (chibenwa) [reply mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/manager/InMemoryIntegrationResources.java]: I found a better way by exposing the knowledge of  `ListeningSearchIndex` through the API of `InMemoryIntegrationResources`

File: mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/manager/InMemoryIntegrationResources.java (hunk 4 above)
Comment (chibenwa) [reply mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/manager/InMemoryIntegrationResources.java]: I did contribute that staged factory.

A convenient `public static InMemoryIntegrationResources defaultResources()` makes the above proposal reasonable.

File: mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/ElasticSearchIntegrationTest.java [hunk 7]
```diff
@@ -113,59 +97,34 @@ protected void initializeMailboxManager() {
...
-
-        eventBus.register(elasticSearchListeningMessageSearchIndex);
-        this.messageSearchIndex = elasticSearchListeningMessageSearchIndex;
+
+        InMemoryIntegrationResources resources = InMemoryIntegrationResources.builder()
+            .preProvisionnedFakeAuthenticator()
+            .fakeAuthorizator()
+            .inVmEventBus()
+            .defaultAnnotationLimits()
+            .defaultMessageParser()            .listeningSearchIndex(preInstanciationStage -> new ElasticSearchListeningMessageSearchIndex(
```
Comment (mbaechler) [inline mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/ElasticSearchIntegrationTest.java]: indent failure

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**